snykTask/src/install/index.ts

```diff
@@ -62,69 +66,62 @@ export async function downloadExecutable(
 ) {
   const filePath = path.join(targetDirectory, executable.filename);
 
-  // Check if the file already exists
   if (fs.existsSync(filePath)) {
     console.log(
       `File ${executable.filename} already exists, skipping download.`,
     );
     return;
   }
 
-  const fileWriter = fs.createWriteStream(filePath, {
-    mode: 0o766,
-  });
-
-  // Wrapping the download in a function for easy retrying
-  const doDownload = () =>
+  const doDownload = (url: string) =>
     new Promise<void>((resolve, reject) => {
-      https
-        .get(executable.downloadUrl, (response) => {
-          const isResponseError = response.statusCode !== 200;
-
-          response.on('error', (err) => {
-            console.error(
-              `Download of ${executable.filename} failed: ${err.message}`,
-            );
-            reject(err);
-          });
+      const fileWriter = fs.createWriteStream(filePath, { mode: 0o766 });
 
+      https
+        .get(url, (response) => {
           if (response.statusCode !== 200) {
             fileWriter.close();
+            reject(new Error(`HTTP ${response.statusCode}`));
+            return;
           }
 
-          fileWriter.on('close', () => {
-            console.log(
-              `File.close ${executable.filename} saved to ${filePath}`,
-            );
-            if (isResponseError) {
-              reject(new Error(`HTTP ${response.statusCode}`));
-            } else {
-              resolve();
-            }
-          });
-
           response.pipe(fileWriter);
+
+          fileWriter.on('finish', () => {
+            fileWriter.close();
+            console.log(`File ${executable.filename} saved to ${filePath}`);
+            resolve();
+          });
         })
         .on('error', (err) => {
-          console.error(
-            `Request for ${executable.filename} failed: ${err.message}`,
-          );
+          fileWriter.close();
+          fs.unlinkSync(filePath); // Delete partially downloaded file
           reject(err);
         });
     });
 
-  // Try to download the file, retry up to `maxRetries` times if the attempt fails
   for (let attempt = 0; attempt < maxRetries; attempt++) {
     try {
-      await doDownload();
+      await doDownload(executable.downloadUrl);
       console.log(`Download successful for ${executable.filename}`);
-      break;
+      return; // Exit function on successful download
     } catch (err) {
       console.error(
-        `Download of ${executable.filename} failed: ${err.message}`,
+        `Download of ${executable.filename} from main URL failed: ${err.message}! Attempting to download from backup URL...`,
       );
 
-      // Don't wait before retrying the last attempt
+      try {
```

Comment:
Question: I'm wondering if we want to use the backup when the first download fails or when max retries failed.  From the perspective that we want to use the main URL, we might want to only use backup if max retires for main failed. Maybe it is even sufficient to move the backup url usage out of the look completely.